pallets/anchors-v2/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		/// The type used to version a document.
+		type DocumentVersion: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		/// Origin used when setting a deposit.
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Weight information.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Storage for document anchors.
+	#[pallet::storage]
+	#[pallet::getter(fn get_anchor)]
+	pub type Anchors<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_256,
+		(T::DocumentId, T::DocumentVersion),
+		Blake2_256,
+		T::AccountId,
+		AnchorOf<T>,
+	>;
+
+	/// Storage for document anchors specific to an account.
+	#[pallet::storage]
+	#[pallet::getter(fn get_personal_anchor)]
+	pub type PersonalAnchors<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_256, T::AccountId>,
+			NMapKey<Blake2_256, T::DocumentId>,
+			NMapKey<Blake2_256, T::DocumentVersion>,
+		),
+		AnchorOf<T>,
+		OptionQuery,
+	>;
+
+	/// Stores the current deposit that will be taken when storing an anchor.
+	#[pallet::storage]
+	#[pallet::getter(fn get_anchor_deposit)]
+	pub type AnchorDeposit<T: Config> =
+		StorageValue<_, T::Balance, ValueQuery, T::DefaultAnchorDeposit>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An anchor was added.
+		AnchorAdded {
+			account_id: T::AccountId,
+			document_id: T::DocumentId,
+			document_version: T::DocumentVersion,
+			hash: T::Hash,
+			deposit: T::Balance,
+		},
+		/// An anchor was removed.
+		AnchorRemoved {
+			account_id: T::AccountId,
+			document_id: T::DocumentId,
+			document_version: T::DocumentVersion,
+			hash: T::Hash,
+			deposit: T::Balance,
+		},
+		/// A deposit was set.
+		DepositSet { new_deposit: T::Balance },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The anchor already exists.
+		AnchorAlreadyExists,
+
+		/// The personal anchor already exists.
+		PersonalAnchorAlreadyExists,
+
+		/// The anchor was not found in storage.
+		AnchorNotFound,
+
+		/// The personal anchor was not found in storage.
+		PersonalAnchorNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sets an anchor for a document ID and version.
+		#[pallet::weight(T::WeightInfo::set_anchor())]
+		#[pallet::call_index(0)]
+		pub fn set_anchor(
+			origin: OriginFor<T>,
+			document_id: T::DocumentId,
+			document_version: T::DocumentVersion,
+			hash: T::Hash,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			// Only one anchor should be stored for a particular document ID and version.
+			ensure!(
+				Anchors::<T>::iter_prefix_values((document_id, document_version)).count() == 0,
```

Comment:
This iterates the whole storage, which could lead to a big weight. 
I think `Anchors::<T>::get((document_id, document_version)).is_none()` do the expected, isn't? As you did below

Reply:
Yup, I also agree that the current approach needs improvement. I can't do the call you suggested since it's a double map, but I'll look into removing the account ID key and try out what you suggested in a different comment.

Reply:
Ahh you're right. Then I think this can be useful: https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.CountedStorageNMap.html

Reply:
I was thinking about this.

If we only allow calling `set_anchor()` if the count is 0, it means that the total entries can only be 0 or 1, no more. Does it means that the storage could be modified to a single `StorageMap`?:

```rust
    pub type Anchors<T: Config> = StorageMap<
		_,
		Blake2_256,
		(T::DocumentId, T::DocumentVersion),
		(T::AccountId, AnchorOf<T>),
	>;
```

Reply:
I updated the storage setup.